rewriter/Minitest.cc

```diff
@@ -12,6 +13,100 @@ using namespace std;
+
+public:
+    unique_ptr<ast::Expression> createConstAssign(ast::Assign &asgn) {
+        auto loc = asgn.loc;
+        auto unsafeNil = ast::MK::Unsafe(loc, ast::MK::Nil(loc));
+        if (auto send = ast::cast_tree<ast::Send>(asgn.rhs.get())) {
+            if (send->fun == core::Names::let() && send->args.size() == 2) {
+                auto rhs = ast::MK::Let(loc, move(unsafeNil), send->args[1]->deepCopy());
+                return ast::MK::Assign(asgn.loc, move(asgn.lhs), move(rhs));
+            }
+        }
+
+        return ast::MK::Assign(asgn.loc, move(asgn.lhs), move(unsafeNil));
+    }
+
+    unique_ptr<ast::Expression> postTransformAssign(core::MutableContext ctx, unique_ptr<ast::Assign> asgn) {
+        if (auto cnst = ast::cast_tree<ast::UnresolvedConstantLit>(asgn->lhs.get())) {
+            if (ast::isa_tree<ast::UnresolvedConstantLit>(asgn->rhs.get())) {
+                movedConstants.emplace_back(move(asgn));
+                return ast::MK::EmptyTree();
+            }
+            auto name = ast::MK::Symbol(cnst->loc, cnst->cnst);
+
+            // if the constant is already in a T.let, preserve it, otherwise decay it to unsafe
+            movedConstants.emplace_back(createConstAssign(*asgn));
+
+            auto module = ast::MK::Constant(asgn->loc, core::Symbols::Module());
+            auto const_set = ctx.state.enterNameUTF8("const_set");
+            return ast::MK::Send2(asgn->loc, move(module), const_set, move(name), move(asgn->rhs));
+        }
+
+        return asgn;
+    }
+
+    // classdefs define new constants, so we always move those if they're the "top-level" classdef (i.e. if we have
+    // nested classdefs, we should only move the outermost one)
+    unique_ptr<ast::ClassDef> preTransformClassDef(core::MutableContext ctx, unique_ptr<ast::ClassDef> classDef) {
+        classDepth++;
+        return classDef;
+    }
+
+    unique_ptr<ast::Expression> postTransformClassDef(core::MutableContext ctx, unique_ptr<ast::ClassDef> classDef) {
+        classDepth--;
+        if (classDepth == 0) {
+            movedConstants.emplace_back(move(classDef));
+            return ast::MK::EmptyTree();
+        }
+        return classDef;
+    }
+
+    // we move sends if they are other minitest `describe` blocks, as those end up being classes anyway: consequently,
+    // we treat those the same wa we treat classes
+    unique_ptr<ast::Send> preTransformSend(core::MutableContext ctx, unique_ptr<ast::Send> send) {
+        if (!send->recv->isSelfReference() && send->args.size() == 1 && send->fun == core::Names::describe()) {
+            classDepth++;
+        }
+        return send;
+    }
+
+    unique_ptr<ast::Expression> postTransformSend(core::MutableContext ctx, unique_ptr<ast::Send> send) {
+        if (!send->recv->isSelfReference() && send->args.size() == 1 && send->fun == core::Names::describe()) {
+            classDepth--;
+            if (classDepth == 0) {
```

Comment:
Resolved this IRL, but this condition fires because the treemap is used not on the content of the whole file, but on subtrees.